src/plugins/bmqauthnbasic/bmqauthnbasic_basicauthenticator.h

```diff
@@ -0,0 +1,186 @@
+
+class BasicAuthenticator : public mqbplug::Authenticator {
+    // CLASS-SCOPE CATEGORY
+    BALL_LOG_SET_CLASS_CATEGORY("MQBAUTHN.BASICAUTHENTICATOR");
+
+  private:
+    // PRIVATE TYPES
+    bsl::vector<char> d_authPayload;
+    bsl::string       d_clientIpAddress;
+
+  private:
+    // DATA
+    const mqbcfg::AuthenticatorPluginConfig* d_authenticatorConfig_p;
+
+    bool d_isStarted;
+
+    bslma::Allocator* d_allocator_p;
+
+  private:
+    // PRIVATE ACCESSORS
+
+  public:
+    // NOT IMPLEMENTED
+    BasicAuthenticator(const BasicAuthenticator& other)            = delete;
+    BasicAuthenticator& operator=(const BasicAuthenticator& other) = delete;
+
+    // CREATORS
+
+    BasicAuthenticator(bslma::Allocator* allocator);
+
+    /// Destructor.
+    ~BasicAuthenticator() BSLS_KEYWORD_OVERRIDE;
+
+    // MANIPULATORS
+
+    /// Return the name of the plugin.
+    bslstl::StringRef name() const BSLS_KEYWORD_OVERRIDE;
+
+    /// Return the authentication mechanism the Authenticator supports.
+    bslstl::StringRef mechanism() const BSLS_KEYWORD_OVERRIDE;
+
+    /// Authenticate using the data provided in the specified `input`.
+    /// - Return `0` on success, and populate the specified `result` with
+    ///   client identity and its remaining lifetime if it has a fixed
+    ///   duration.
+    /// - Return a non-zero plugin-specific return code upon failure, and
+    ///   populate the specified `errorDescription` with a brief reason for
+    ///   logging purposes.
+    int authenticate(bsl::ostream&                      errorDescript,
+                     mqbplug::AuthenticationResult*     result,
+                     const mqbplug::AuthenticationData& input) const
+        BSLS_KEYWORD_OVERRIDE;
+
+    // MANIPULATORS
+
+    /// Start the Authenticator and return 0 on success, or return a non-zero
+    /// value and populate the specified `errorDescription` with the
+    /// description of any failure encountered.
+    int start(bsl::ostream& errorDescription) BSLS_KEYWORD_OVERRIDE;
+
+    /// Stop the Authenticator.
+    void stop() BSLS_KEYWORD_OVERRIDE;
+};
+
+// =====================================
+// class BasicAuthenticatorPluginFactory
+// =====================================
+
+class BasicAuthenticatorPluginFactory
+: public mqbplug::AuthenticatorPluginFactory {
+  public:
+    // CREATORS
+    BasicAuthenticatorPluginFactory();
+    ~BasicAuthenticatorPluginFactory() BSLS_KEYWORD_OVERRIDE;
+
+    // MANIPULATORS
+    bslma::ManagedPtr<mqbplug::Authenticator>
+    create(bslma::Allocator* allocator) BSLS_KEYWORD_OVERRIDE;
+};
+
+// ============================================================================
+//                             INLINE DEFINITIONS
+// ============================================================================
+
+// -----------------------------
+// class BasicAuthenticationData
+// -----------------------------
+
+inline const bsl::vector<char>& BasicAuthenticationData::authPayload() const
```

Comment:
Virtual methods can't be declared inline; this will result in the vtable being emitted in each translation unit, which in this case will likely cause an issue when the plugin is loaded.